velox/functions/prestosql/Sequence.cpp

```diff
@@ -0,0 +1,173 @@
+      totalCount += count;
+    });
+
+    VectorPtr elements = BaseVector::create(BIGINT(), totalCount, pool);
+    BufferPtr sizes = allocateSizes(numRows, pool);
+    BufferPtr offsets = allocateOffsets(numRows, pool);
+    auto rawElements = elements->asFlatVector<int64_t>()->mutableRawValues();
+    auto rawSizes = sizes->asMutable<vector_size_t>();
+    auto rawOffsets = offsets->asMutable<vector_size_t>();
+
+    int elementsOffset = 0;
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      rawOffsets[row] = elementsOffset;
+      rawSizes[row] = sequenceCount(startVector, stopVector, stepVector, row);
+      writeToElements(
+          rawElements,
+          elementsOffset,
+          startVector,
+          stopVector,
+          stepVector,
+          row);
+      elementsOffset += rawSizes[row];
+    });
+
+    return std::make_shared<ArrayVector>(
+        pool, outputType, nullptr, numRows, offsets, sizes, elements);
+  }
+
+  int sequenceCount(
+      VectorPtr startVector,
+      VectorPtr stopVector,
+      VectorPtr stepVector,
+      vector_size_t row) const {
+    long start = valueInVector(startVector, row);
+    long stop = valueInVector(stopVector, row);
+    long step;
+    if (stepVector == nullptr) {
+      step = stop >= start ? 1 : -1;
+    } else {
+      step = valueInVector(stepVector, row);
+    }
+    checkArguments(start, stop, step);
+    return (int)((stop - start) / step + 1);
+  }
+
+  long valueInVector(VectorPtr baseVector, vector_size_t row) const {
+    if (baseVector->isConstantEncoding()) {
+      return baseVector->as<ConstantVector<long long>>()->valueAt(0);
+    } else {
+      return baseVector->asFlatVector<long long>()->valueAt(row);
+    }
+  }
+
+  void checkArguments(int start, int stop, int step) const {
+    VELOX_USER_CHECK_NE(step, 0, "step must not be zero");
+    VELOX_USER_CHECK(
+        step > 0 ? stop >= start : stop <= start,
+        "sequence stop value should be greater than or equal to start value if "
+        "step is greater than zero otherwise stop should be less than or equal to start")
+    VELOX_USER_CHECK_LE(
+        (stop - start) / step + 1,
+        kMaxResultEntries,
+        "result of sequence function must not have more than 10000 entries");
+  }
+
+  void writeToElements(
+      long long* elements,
+      int elementsOffset,
+      VectorPtr startVector,
+      VectorPtr stopVector,
+      VectorPtr stepVector,
+      vector_size_t row) const {
+    long start = valueInVector(startVector, row);
+    long stop = valueInVector(stopVector, row);
+    long step;
+    if (stepVector == nullptr) {
+      step = stop >= start ? 1 : -1;
+    } else {
+      step = valueInVector(stepVector, row);
+    }
+
+    long length = (stop - start) / step + 1;
+    for (long i = 0, value = start; i < length; ++i, value += step) {
+      elements[elementsOffset] = value;
+      elementsOffset++;
+    }
```

Comment:
Would this be a little bit more clear:
```
const auto numElements = (stop - start) / step + 1;
const auto valueEnd = start + stop * numElements;
for (auto value = start, value != valueEnd; value += step) {
  elements[elementsOffset++] = value;
}
```
?

Also this function is quite simple - it might be better to just leave its body in the context.applyToSelectedNoThrow's lambda.

Reply:
Just changed the logic
But I haven't moved the logic into the lambda, because from the java Sequence function, this function could be used in multiple places (https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/operator/scalar/SequenceFunction.java#L147), so I think we would be reuse this function soon when we add more support